modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryTupleParser.java

```diff
@@ -0,0 +1,215 @@
+        /**
+         * Provides the location of the next tuple value.
+         *
+         * @param index Value index.
+         * @param begin Start offset of the value, 0 for NULL.
+         * @param end End offset of the value, 0 for NULL.
+         */
+        void nextElement(int index, int begin, int end);
+    }
+
+    /** Number of elements in the tuple. */
+    private final int numElements;
+
+    /** Size of an offset table entry. */
+    private final int entrySize;
+
+    /** Position of the varlen offset table. */
+    private final int entryBase;
+
+    /** Starting position of variable-length values. */
+    private final int valueBase;
+
+    /** Binary tuple. */
+    protected final ByteBuffer buffer;
+
+    /**
+     * Constructor.
+     *
+     * @param numElements Number of tuple elements.
+     * @param buffer Buffer with a binary tuple.
+     */
+    public BinaryTupleParser(int numElements, ByteBuffer buffer) {
+        this.numElements = numElements;
+
+        assert buffer.order() == ByteOrder.LITTLE_ENDIAN;
+        assert buffer.position() == 0;
+        this.buffer = buffer;
+
+        byte flags = buffer.get(0);
+
+        int base = BinaryTupleSchema.HEADER_SIZE;
+        if ((flags & BinaryTupleSchema.NULLMAP_FLAG) != 0) {
+            base += BinaryTupleSchema.nullMapSize(numElements);
+        }
+
+        entryBase = base;
+        entrySize = 1 << (flags & BinaryTupleSchema.VARSIZE_MASK);
+        valueBase = base + entrySize * numElements;
+    }
+
+    /**
+     * Returns the binary tuple size in bytes.
+     */
+    public int size() {
+        return valueBase + getOffset(valueBase - entrySize);
+    }
+
+    /**
+     * Returns the number of elements in the tuple.
+     */
+    public int elementCount() {
+        return numElements;
+    }
+
+    /**
+     * Check if the binary tuple contains a null map.
+     */
+    public boolean hasNullMap() {
+        return entryBase > BinaryTupleSchema.HEADER_SIZE;
+    }
+
+    /**
+     * Get buffer positioned at the specified tuple element.
+     *
+     * If the element has NULL value then null is returned.
+     *
+     * @param index Index of the element.
+     * @return ByteBuffer with element bytes between current position and limit or null.
+     */
+    public ByteBuffer element(int index) {
+        var sink = new Sink() {
+            int begin, end;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                this.begin = begin;
+                this.end = end;
+            }
+        };
+
+        fetch(index, sink);
+        if (sink.begin == 0) {
+            return null;
+        }
+
+        return buffer.asReadOnlyBuffer().position(sink.begin).limit(sink.end).order(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    /**
+     * Locate the specified tuple element.
+     *
+     * @param index Index of the element.
+     * @param sink Receiver.
+     */
+    public void fetch(int index, Sink sink) {
```

Comment:
Sink param is used here to pass more than one parameter to the outsize the method.
Why can't it be a return type?